### PR TITLE
Add more guards against threading misuse and races.

### PR DESCRIPTION
**Pull request by MikeStall**
Add more guards against threading misuse and races.
No functional changes.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs [hunk 1]
```diff
@@ -34,24 +34,32 @@ internal class TexlFunctionSet
 
         internal Dictionary<DPath, List<TexlFunction>>.KeyCollection Namespaces => _namespaces.Keys;
 
-        internal IEnumerable<TexlFunction> WithName(string name) => WithNameInternal(name);
+        internal IEnumerable<TexlFunction> WithName(string name) => _guard.VerifyNoWriters(this).WithNameInternal(name);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Functions/TexlFunctionSet.cs:37]: These are key asserts - existing guards catch 2 writers. 
This check is for 1 reader during writers.

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs [hunk 2]
```diff
@@ -338,17 +338,13 @@ public SymbolTable GetMutableCopyOfFunctions()
...
 
-            s.AddFunctions(_functions);
+            s.AddFunctions(this.Functions);
 
             return s;
         }
 
         internal readonly Dictionary<string, NameLookupInfo> _variables = new Dictionary<string, NameLookupInfo>();
-
-        private protected readonly TexlFunctionSet _functions = new TexlFunctionSet();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs:348]: Move the mutable one out of the base class and into derived SymbolTable class. 
This lets the derived class assert guards on it.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.ReadOnlySymbolTable.get_FunctionNames

File: src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs [hunk 3]
```diff
@@ -61,29 +61,35 @@ public override FormulaType GetTypeFromSlot(ISymbolSlot slot)
...
+
         // Expose the list to aide in intellisense suggestions. 
         // Multiple readers ok. But not writing while we read.
         TexlFunctionSet INameResolver.Functions
         {
             get
             {
-                var current = this.VersionHash;
-                if (current != _cachedVersionHash)
+                lock (_lock)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Config/ComposedReadOnlySymbolTable.cs:73]: @jas-valgotar  - fyi. 
This lock doesn't prevent mutations. but it should ensure the fields are updated consistently and avoid the null refs.

**Review by jas-valgotar — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**